Userland/useradd.cpp

```diff
@@ -0,0 +1,110 @@
+#include <AK/String.h>
+#include <AK/StringBuilder.h>
+#include <LibCore/CArgsParser.h>
+#include <pwd.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+static constexpr uid_t BASE_UID = 1000;
+static constexpr gid_t USERS_GID = 100;
+const char* DEFAULT_SHELL = "/bin/sh";
+
+int main(int argc, char** argv)
+{
+    CArgsParser args_parser("useradd");
+    args_parser.add_arg("d", "HOME_DIR", "home directory for the new user");
+    args_parser.add_arg("u", "UID", "user ID (uid) for the new user");
+    args_parser.add_arg("g", "GROUP", "group ID (gid) for the new user");
+    args_parser.add_arg("m", "create home directory if it does not exist");
+    args_parser.add_arg("s", "SHELL", "name of default shell binary for the new user");
+    args_parser.add_arg("n", "NAME", "gecos name of the new user");
+    CArgsParserResult args = args_parser.parse(argc, argv);
+
+    if (argc <= 1) {
+        args_parser.print_usage();
+        return 1;
+    }
+
+    const char* username = argv[argc - 1]; // Always the user name!
+    uid_t uid = BASE_UID;
+    gid_t gid = USERS_GID;
+
+    // First, let's sort out the uid for the user
+    if (args.is_present("u")) {
+        bool ok;
+        uid = static_cast<uid_t>(args.get("u").to_uint(ok));
+
+        if (getpwuid(uid)) {
+            fprintf(stderr, "uid %u already exists!\n", uid);
+            return 4;
+        }
+
+        if (!ok) {
+            fprintf(stderr, "Invalid uid %s\n", args.get("u").characters());
+            return -1;
+        }
+    } else {
+        uid = BASE_UID;
+        for (; getpwuid(uid); uid++) {
+        }
+    }
+
+    // Now let's sort out the gid!
+    if (args.is_present("g")) {
+        bool ok;
+        gid = static_cast<gid_t>(args.get("g").to_uint(ok));
+
+        if (!ok) {
+            fprintf(stderr, "Invalid gid %s\n", args.get("g").characters());
+            return -1;
+        }
+    }
+
+    AK::String shell = args.is_present("s") ? args.get("s").characters() : DEFAULT_SHELL;
+    AK::String home = args.is_present("d") ? args.get("d").characters() : "/";
+    AK::String gecos = args.is_present("n") ? args.get("n").characters() : "Anonymous";
+
+    if (args.is_present("m")) {
+        AK::StringBuilder builder;
+        builder.append("/home/");
+        builder.append(username);
+
+        AK::String home_path(builder.to_string());
+        if (mkdir(home_path.characters(), 0700)) {
+            perror("mkdir");
+            return 12;
+        }
+
+        if (chown(home_path.characters(), uid, gid)) {
+            perror("chown");
+            return 12;
+        }
+
+        home = home_path.characters(); // This overwrites the -d argument
+    }
+
+    struct passwd p;
+    p.pw_name = const_cast<char*>(username);
+    p.pw_dir = const_cast<char*>(home.characters());
+    p.pw_uid = uid;
+    p.pw_gid = gid;
+    p.pw_shell = const_cast<char*>(shell.characters());
```

Comment:
...or maybe you only need a C string here; then do use `.characters()` and store the variable as a raw char pointer

Reply:
Might keep it as a string for now considering all the issues I was getting with the conversion. I'll refactor it later :P